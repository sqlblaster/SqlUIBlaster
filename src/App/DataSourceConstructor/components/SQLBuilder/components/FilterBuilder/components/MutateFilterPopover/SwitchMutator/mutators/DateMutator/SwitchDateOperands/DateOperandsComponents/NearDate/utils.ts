import { PluralDateComponent } from './plural-date-components';

export const getThisDateComponentNotation = (
  dateComponent: PluralDateComponent
) => {
  const singularForm = dateComponent.substr(0, dateComponent.length - 1);

  const thisDateNotation =
    singularForm === 'Day' ? 'Today' : `This ${singularForm}`;

  return thisDateNotation.toLowerCase();
};
