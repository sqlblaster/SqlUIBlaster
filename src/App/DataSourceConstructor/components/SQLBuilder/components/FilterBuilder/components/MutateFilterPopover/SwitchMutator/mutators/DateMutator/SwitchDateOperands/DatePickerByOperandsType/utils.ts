import { SingleDateOperands } from '../DateOperandsComponents/SingleDate/model';
import { FixedDateOperands } from './components/FixedDatePicker/model';
import { RelativeDateOperands } from './components/RelativeDatePicker/model';
import { VariableDateOperands } from './components/VariableDatePicker/model';
import { DateType } from './date-operand-types';

export const getNewOperands = (dateType: DateType): SingleDateOperands => {
  switch (dateType) {
    case 'Fixed date':
      return new FixedDateOperands();
    case 'Relative date':
      return new RelativeDateOperands();
    case 'Variable date':
      return new VariableDateOperands();
    default:
      throw new Error('unknown date type is provided');
  }
};
