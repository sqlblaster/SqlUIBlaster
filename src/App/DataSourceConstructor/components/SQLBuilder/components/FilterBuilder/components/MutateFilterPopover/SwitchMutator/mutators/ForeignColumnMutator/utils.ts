import { ForeignColumnMutatorProps } from './props';

export const getModels = (
  props: Pick<ForeignColumnMutatorProps, 'enums' | 'column'>
): string[] | undefined => {
  const {
    enums,
    column: {
      lastColumn: { foreignModelName, modelName, type }
    }
  } = props;

  return type === 'uuid' && !enums
    ? undefined
    : type === 'uuid' && enums
    ? enums
    : type === 'primary'
    ? [modelName]
    : foreignModelName
    ? [foreignModelName]
    : undefined;
};
