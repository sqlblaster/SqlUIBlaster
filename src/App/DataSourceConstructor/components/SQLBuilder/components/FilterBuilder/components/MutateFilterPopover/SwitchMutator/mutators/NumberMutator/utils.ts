import { NumberOperands } from './NumberOperation';
import { NumberOperator } from './operators';

export const getDefaultOperands = (
  operator: NumberOperator
): NumberOperands => {
  if (operator === 'Between') {
    return [0, 0];
  } else if (operator === 'Is empty' || operator === 'Is not empty') {
    return undefined;
  } else {
    return 0;
  }
};
