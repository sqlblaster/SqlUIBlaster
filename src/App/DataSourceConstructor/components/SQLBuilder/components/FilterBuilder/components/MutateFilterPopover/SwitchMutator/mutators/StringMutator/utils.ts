import { StringOperator } from './operators';

export const hasCaseSensivity = (operator: StringOperator): boolean => {
  switch (operator) {
    case 'Contains':
    case 'Does not contain':
    case 'Starts with':
    case 'Ends with':
      return true;
    default:
      return false;
  }
};
