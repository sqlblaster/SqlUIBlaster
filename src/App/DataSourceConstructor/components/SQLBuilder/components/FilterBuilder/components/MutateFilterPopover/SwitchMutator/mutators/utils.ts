import {
  Column,
  Table
} from 'src/App/DataSourceConstructor/schemas';
import { ColumnBranch } from '../../../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';

export const getEnums = (
  tables: Table[],
  columnBranch: ColumnBranch
): Column['enums'] => {
  const { tableName, columnName } = columnBranch;
  let enums: Column['enums'];

  const table = tables.find(t => t.name === tableName);
  let column: Column | undefined;
  table && (column = table.columns.find(c => c.name === columnName));
  column && (enums = column.enums);

  return enums;
};
