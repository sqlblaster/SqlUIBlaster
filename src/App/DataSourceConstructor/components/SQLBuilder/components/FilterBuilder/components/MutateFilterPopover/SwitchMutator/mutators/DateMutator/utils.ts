import { DateOperation } from './DateOperation';
import { CoupleDateOperation } from './SwitchDateOperands/DateOperandsComponents/CoupleDates/model';
import { EmptyDateOperation } from './SwitchDateOperands/DateOperandsComponents/EmptyDates/model';
import { NearDateOperation } from './SwitchDateOperands/DateOperandsComponents/NearDate/model';
import { SingleDateOperation } from './SwitchDateOperands/DateOperandsComponents/SingleDate/model';

export const getDateOperationVia = (
  operator: DateOperation['operator']
): DateOperation => {
  switch (operator) {
    case 'Previous':
    case 'Next':
      return new NearDateOperation({ operator });
    case 'Before':
    case 'After':
    case 'On':
      return new SingleDateOperation({ operator });
    case 'Between':
      return new CoupleDateOperation({ operator });
    case 'Is empty':
    case 'Is not empty':
    default:
      return new EmptyDateOperation({ operator });
  }
};
