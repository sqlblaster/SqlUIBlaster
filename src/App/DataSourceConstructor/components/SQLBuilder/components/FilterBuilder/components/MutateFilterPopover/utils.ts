import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { Operation } from './Operation';
import { BooleanOperation } from './SwitchMutator/mutators/BooleanMutator/BooleanOperation';
import { NearDateOperation } from './SwitchMutator/mutators/DateMutator/SwitchDateOperands/DateOperandsComponents/NearDate/model';
import { ForeignColumnOperation } from './SwitchMutator/mutators/ForeignColumnMutator/ForeignColumnOperation';
import { NumberOperation } from './SwitchMutator/mutators/NumberMutator/NumberOperation';
import { StringOperation } from './SwitchMutator/mutators/StringMutator/StringOperation';

export const getDefaultOperation = (columnBranch: ColumnBranch): Operation => {
  if (columnBranch.foreignTableName) {
    return new ForeignColumnOperation();
  }

  switch (columnBranch.type) {
    case 'number':
      return new NumberOperation();
    case 'string':
      return new StringOperation();
    case 'boolean':
      return new BooleanOperation();
    case 'Date':
      return new NearDateOperation();
    default:
      return new StringOperation();
  }
};
