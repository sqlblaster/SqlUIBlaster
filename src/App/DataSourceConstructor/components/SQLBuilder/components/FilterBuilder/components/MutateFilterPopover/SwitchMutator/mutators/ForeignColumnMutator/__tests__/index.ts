import { fireEvent, getByText as getByTextOn } from 'react-testing-library';
import { variableIconPlaceholder } from 'src/__tests__/utils/mock-placeholders';
import {
  getGeneratedQuery,
  render
} from 'src/App/DataSourceConstructor/components/SQLBuilder/__tests__/render.utils';
import {
  clickAddTheFilter,
  clickUpdateTheFilter
} from 'src/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/__tests__/utils/filter-utils';
import { selectATable } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/TableSelector/__tests__/utils/select-table';
import {
  fireChangeEvent,
  getByText$,
  regex
} from 'src/App/DataSourceConstructor/test-utils/utils';
import { foreignFieldPlaceholder } from 'src/process/startnew/ForeignField';
import { clickAddAFilter } from '../../../../../AddFilter/__tests__/utils/click-add-a-filter';
import {
  addFilterText,
  mutatorPanelId,
  updateFilterText
} from '../../../../view';
import { foreignFieldDataMock } from './utils/foreign-field-data.mock';
import { foreignFieldVariablesMock } from './utils/foreign-field-variables.mock';

jest.mock('src/utils');

jest.mock('react-apollo', () => {
  const {
    foreignFieldDataMock: dataMock
    // tslint:disable-next-line: no-require-imports
  } = require('./utils/foreign-field-data.mock');
  const {
    getConfiguredQuery,
    getResult
  // tslint:disable-next-line: no-require-imports
  } = require('src/__dynamic_mocks__/react-apollo.Query.mock');

  return {
    Query: getConfiguredQuery(() =>
      getResult({
        data: dataMock
      })
    )
  };
});

describe(`Foreign field operations`, () => {
  it(`should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection`, async () => {
    const {
      getByText,
      getByTestId,
      getByPlaceholderText,
      rerender,
      componentNode
    } = render();

    const recordsMock = foreignFieldDataMock.navigation.search.records;

    const columns = selectATable();
    const foreignColumnName = columns[6].name;

    const clickOnFilter = () => {
      fireEvent.click(getByText(regex(foreignColumnName)));
    };

    const clickForeignColumn = () => {
      fireEvent.click(
        getByTextOn(getByTestId(mutatorPanelId), foreignColumnName)
      );
    };

    const checkFilterDescription = (snapName: string) => {
      expect(
        getByText$(componentNode, new RegExp(`^${foreignColumnName}.+$`, 'i'))
          .textContent
      ).toMatchSnapshot(`${snapName} filter description`);
    };

    // an item selected
    clickAddAFilter();
    clickForeignColumn();

    expect(getByText(addFilterText)).toBeDisabled();

    let input = getByPlaceholderText(foreignFieldPlaceholder);
    fireEvent.focus(input);
    fireChangeEvent(input, recordsMock[0].title);
    fireEvent.click(getByText(recordsMock[0].title));

    expect(getByText(addFilterText)).toBeEnabled();

    clickAddTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot(`an item selected`);

    checkFilterDescription('an item selected');

    // a variable selected after an item
    rerender({ variables: foreignFieldVariablesMock });

    clickOnFilter();

    expect(getByText(updateFilterText)).toBeEnabled();

    fireEvent.click(getByText(variableIconPlaceholder));
    fireEvent.click(getByText(new RegExp(foreignFieldVariablesMock[0].name)));

    clickUpdateTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot(
      `a variable selected after an item`
    );

    checkFilterDescription('a variable selected after an item');

    // an item selected after variable
    clickOnFilter();

    expect(getByText(updateFilterText)).toBeEnabled();

    input = getByPlaceholderText(foreignFieldPlaceholder);
    fireEvent.focus(input);
    fireChangeEvent(input, recordsMock[1].title);
    fireEvent.click(getByText(recordsMock[1].title));

    clickUpdateTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot(
      `an item selected after variable`
    );

    checkFilterDescription('an item selected after variable');
  });

  it.skip(`
      field with 'GraphNode' model should work the same way as foreign field and 'primary' field in the foreign table,
      'created_by', 'updated_by', 'ownerid' fields should have 'User' model and 'User' variables
      'uuid' field should provide all models data and has all Record Variables,
      'polymorph' field should have only certain models
      'polymorph' field should have certain variables list which corresponds to its models list
      'polyenum' field should have enums list
      'primary' type field should have 'key' icon
      'primary' type field in the foreign table should work the same way as selected table's 'primary' type field
  `, () => {
    // empty block
  });
});
