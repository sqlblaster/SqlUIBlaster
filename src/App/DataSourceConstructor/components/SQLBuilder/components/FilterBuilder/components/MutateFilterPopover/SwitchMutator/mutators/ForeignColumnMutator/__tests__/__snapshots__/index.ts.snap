// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Foreign field operations should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection: a variable selected after an item 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"UserID\\" = $1::uuid
LIMIT 2000
"
`;

exports[`Foreign field operations should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection: a variable selected after an item filter description 1`] = `"UserIDisVariableIconUSER"`;

exports[`Foreign field operations should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection: an item selected 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"UserID\\" = '1'
LIMIT 2000
"
`;

exports[`Foreign field operations should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection: an item selected after variable 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"UserID\\" = '2'
LIMIT 2000
"
`;

exports[`Foreign field operations should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection: an item selected after variable filter description 1`] = `"UserIDisKevin"`;

exports[`Foreign field operations should render proper sql query on:
      - an item selected
      - a variable selected after an item
      - an item selected after variable,
      should not allow save filter on empty selection: an item selected filter description 1`] = `"UserIDisDavid"`;
